### PR TITLE
Add support for Blade Pro 2016 keyboard

### PR DESCRIPTION
**Pull request by hyperb1iss**
* Uses a similar configuration as the Stealth 2016 with a few
   extra features.

### CODE REVIEW

**Comment by hyperb1iss** [comment]
Updated with various fixes and rebased. The only issue is with udev. For this model, udev needs to match on bInterfaceNumber==2 or chaos is the result. I noticed that Ubuntu does this already and your README has a udev note in it as well, so I left it alone for now.

**Comment by terricain** [comment]
What happens if udev binds all interfaces?

**Comment by hyperb1iss** [comment]
A bunch of F keys stop working, the media keys die, the wheel dies, and the FN key prints G.

**Comment by terricain** [comment]
Hmm that'll be that evil remapping function. Ok just to keep it consistent with the rest I think it would be better to just put an if statement inside the remapping functions to skip the razer blades. That way all interfaces can be bound the same as the rest of the keyboards.

**Comment by hyperb1iss** [comment]
As in, skip the find_translation stuff? Will give that a try.

**Comment by terricain** [comment]
Yeah exactly. You should only have to skip stuff in the `razer_event` function though you might also have to do the same to `razer_raw_event` as well (though in theory that shouldn't do anything, that or its the main culprit).

**Comment by hyperb1iss** [comment]
It's definitely razer_raw_event. Skipping the translation in razer_event does nothing, but returning 0 from razer_raw_event changes the behavior- the FN key doesn't print G anymore and all F-key work but the media keys and wheel report nothing instead of wrong keycodes. The Pro has a different layout than all the other keyboards, I think. Instead of a numeric keypad, they put the trackpad to the *right* of the keyboard with 4 media keys and a wheel (bound to vol up/down by default) above the trackpad. Will try to debug it.

**Comment by terricain** [comment]
Yeah raw_event was a quality hack to map some of the dodgy macro button keycodes to normal keycodes

**Comment by hyperb1iss** [comment]
This is odd, I think there may be some shenanigans going on here. Returning zero from event/raw_event should make it behave exactly like the hid-generic driver (with the sysfs goodies added) unless I'm missing something. I'm suspicious that something is derpy during the driver rebind, since I get a big fat WARN from the USB stack when the razerkbd driver is probed. Seen anything like this?

[    6.048432] ------------[ cut here ]------------
[    6.048435] WARNING: CPU: 4 PID: 628 at /home/kernel/COD/linux/drivers/usb/core/hcd.c:1584 usb_hcd_map_urb_for_dma+0x37f/0x570
[    6.048436] transfer buffer not dma capable
[    6.048495] CPU: 4 PID: 628 Comm: razer_mount Tainted: P           OE   4.9.0-040900rc7-generic #201611271731
[    6.048495] Hardware name: Razer Blade Pro/Blade Pro, BIOS 04.00 11/07/2016
[    6.048496]  ffffad0983fd7800 ffffffffa9417612 ffffad0983fd7850 0000000000000000
[    6.048498]  ffffad0983fd7840 ffffffffa9083d4b 00000630fed193a8 ffffa0cfd8f790c0
[    6.048499]  0000000000000000 ffffa0cfd152e000 0000000000000002 ffffa0cfdaab4000
[    6.048501] Call Trace:
[    6.048503]  [<ffffffffa9417612>] dump_stack+0x63/0x81
[    6.048505]  [<ffffffffa9083d4b>] __warn+0xcb/0xf0
[    6.048507]  [<ffffffffa9083dcf>] warn_slowpath_fmt+0x5f/0x80
[    6.048508]  [<ffffffffa965dd1f>] usb_hcd_map_urb_for_dma+0x37f/0x570
[    6.048509]  [<ffffffffa965f1ec>] usb_hcd_submit_urb+0x34c/0xb40
[    6.048510]  [<ffffffffa902c74e>] ? __switch_to+0x2ce/0x6c0
[    6.048512]  [<ffffffffa988845b>] ? __schedule+0x23b/0x6e0
[    6.048513]  [<ffffffffa9660b7f>] usb_submit_urb.part.6+0x2ff/0x530
[    6.048514]  [<ffffffffa9660e12>] usb_submit_urb+0x62/0x70
[    6.048515]  [<ffffffffa966153e>] usb_start_wait_urb+0x6e/0x170
[    6.048516]  [<ffffffffa966171c>] usb_control_msg+0xdc/0x120
[    6.048519]  [<ffffffffc12b3022>] razer_get_usb_response+0xa2/0xe0 [razerkbd]
[    6.048520]  [<ffffffffc12b06be>] razer_send_payload+0x5e/0x1d0 [razerkbd]
[    6.048522]  [<ffffffffc12b284c>] razer_set_device_mode+0x5c/0x80 [razerkbd]
[    6.048523]  [<ffffffffa92bc903>] ? kernfs_activate+0x83/0xe0
[    6.048524]  [<ffffffffa92bca40>] ? kernfs_add_one+0xe0/0x130
[    6.048526]  [<ffffffffa92be22f>] ? __kernfs_create_file+0x6f/0x90
[    6.048527]  [<ffffffffc12b2b1d>] razer_kbd_probe+0x2ad/0x640 [razerkbd]
[    6.048530]  [<ffffffffc0114597>] hid_device_probe+0xd7/0x150 [hid]
[    6.048531]  [<ffffffffa9594c73>] driver_probe_device+0x223/0x430
[    6.048532]  [<ffffffffa9593054>] bind_store+0x114/0x160
[    6.048533]  [<ffffffffa9592375>] drv_attr_store+0x25/0x30
[    6.048535]  [<ffffffffa92be557>] sysfs_kf_write+0x37/0x40
[    6.048536]  [<ffffffffa92bd89c>] kernfs_fop_write+0x12c/0x1c0
[    6.048537]  [<ffffffffa9236937>] __vfs_write+0x37/0x160
[    6.048539]  [<ffffffffa93b93d8>] ? apparmor_file_permission+0x18/0x20
[    6.048540]  [<ffffffffa937809b>] ? security_file_permission+0x3b/0xc0
[    6.048541]  [<ffffffffa9237065>] vfs_write+0xb5/0x1a0
[    6.048542]  [<ffffffffa92384c5>] SyS_write+0x55/0xc0
[    6.048543]  [<ffffffffa988d67b>] entry_SYSCALL_64_fastpath+0x1e/0xad
[    6.048543] ---[ end trace 9ed9afb760e123f4 ]---

**Comment by terricain** [comment]
Nope but its most likely [this](https://github.com/terrycain/razer-drivers/blob/master/driver/razerkbd_driver.c#L1355) line causing the issue. Try commenting that out and see if it makes a difference. If so then surround that with if statement to exclude it if razer blade.

**Comment by hyperb1iss** [comment]
That line was one issue, but the write from the daemon to switch it to 3 threw up a warn. Also, get_firmware_version fails as well. I stubbed them out and everything works perfect. The daemon fails to start if I simply remove the sysfs entries, should I fix that or just do nothing in those functions?

**Comment by terricain** [comment]
Hmm. Well get fw version we can just return 'v0.0' (would rather be consistent even if we return rubbish). I swore I told the daemon to only switch chroma to 3. Best thing to do would be to exclude the laptops from changing mode as they obviously don't support it :wink:

I'll have a look at the daemon once im off the train. I don't mind patching all of this, but feel free if you want to.

**Comment by hyperb1iss** [comment]
Yeah, what I have is pretty similar already. Will update the patch shortly.

**Comment by hyperb1iss** [comment]
New patch is working 100% for me when bound to all subdevices.

Now to get this stupid Glidepoint trackpad to work as a trackpad instead of a mouse and this thing will be solid :)

**Comment by hyperb1iss** [comment]
One more issue.. read_set_brightness throws a kernel warning, let me see if I can fix.

**Comment by hyperb1iss** [comment]
These kernel warnings about transfer buffers not being DMA capable is a 4.9 specific issue. USB transfer buffers can't be allocated on the stack in 4.9. Will do a separate patch for this.

**Comment by hyperb1iss** [comment]
Finally golden now. First patch fixes the response buffer issue. I was able to re-enable get_firmware_version with this as well (not driver mode though, which kills the media keys).

**Comment by terricain** [comment]
Awesome. Will look properly when I've had sleep :wink:

**Comment by hyperb1iss** [comment]
Rebased and cleaned up the udev script, as it was failing on Ubuntu 16.10.

**Comment by terricain** [comment]
Was it that real hacky udev mount script that was the issue? Nice job on that btw.

Yeah device_mode at least for keyboards sets the kb in a really dumb state where the "driver" in razer words aka razer synapse handles all of the functions including media keys. Thats what the bulk of the nasty event stuff is for, remapping Fn+keys to media and dealing with M1-5.

**Comment by hyperb1iss** [comment]
The script worked, it just didn't always work when udev ran it, and the error message was useless.

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -1367,8 +1412,11 @@ static int razer_kbd_probe(struct hid_device *hdev, const struct hid_device_id *
         goto exit_free;
     }
 
+    // Leave autosuspend on for laptops
+    if (usb_dev->descriptor.idProduct != USB_DEVICE_ID_RAZER_BLADE_PRO_LATE_2016) {
```
Comment (terricain) [inline driver/razerkbd_driver.c]: Would you mind sticking the other razer laptops here

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -1183,7 +1209,14 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
 {
     struct usb_interface *intf = to_usb_interface(hdev->dev.parent);
     struct razer_kbd_device *asc = hid_get_drvdata(hdev);
-    
+    struct usb_device *usb_dev = interface_to_usbdev(intf);
+
+    // No translations needed on the Pro
+    if (usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLADE_PRO_LATE_2016)
```
Comment (terricain) [inline driver/razerkbd_driver.c]: Would you mind sticking the other razer laptops here

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -1125,10 +1143,18 @@ static DEVICE_ATTR(matrix_custom_frame,     0220, NULL,
 static int razer_event(struct hid_device *hdev, struct hid_field *field, struct hid_usage *usage, __s32 value)
 {
     struct usb_interface *intf = to_usb_interface(hdev->dev.parent);
+    struct usb_device *usb_dev = interface_to_usbdev(intf);
     struct razer_kbd_device *asc = hid_get_drvdata(hdev);
     const struct razer_key_translation *translation;
     int do_translate = 0;
 
+    // No translations needed on the Pro
+    if (usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLADE_PRO_LATE_2016)
```
Comment (terricain) [inline driver/razerkbd_driver.c]: Would you mind sticking the other razer laptops here

File: driver/razerkbd_driver.c [hunk 4]
```diff
@@ -979,6 +990,11 @@ static ssize_t razer_attr_write_device_mode(struct device *dev, struct device_at
     struct usb_device *usb_dev = interface_to_usbdev(intf);
     struct razer_report report;
     
+    if (usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLADE_PRO_LATE_2016) {
```
Comment (terricain) [inline driver/razerkbd_driver.c]: Would you mind sticking the other razer laptops here

File: driver/razerkbd_driver.c [hunk 5]
```diff
@@ -150,6 +150,8 @@ void razer_set_device_mode(struct usb_device *usb_dev, unsigned char mode, unsig
         case USB_DEVICE_ID_RAZER_ORNATA_CHROMA:
 			report.transaction_id.id = 0x3F;
 			break;
+        case USB_DEVICE_ID_RAZER_BLADE_PRO_LATE_2016:
```
Comment (terricain) [inline driver/razerkbd_driver.c]: Would you mind sticking the other razer laptops here

**Review by terricain — CHANGES_REQUESTED** [review]
Thought I'd try out this github review thing. Mind adding the other blades to the list of ones that are skipped for the event things, and autosuspend etc...

**Comment by terricain** [comment]
Yeah that script wasnt the best ;)

**Comment by hyperb1iss** [comment]
Sure thing, as long as you are able to test them :)

**Comment by terricain** [comment]
The community will test them :+1: To be fair it wont break them if anything it'll make it better as they were tested long before the event stuff was conceived.

**Comment by hyperb1iss** [comment]
Done, and did some code consolidation too.

**Review by terricain — APPROVED** [review]

**Comment by terricain** [comment]
Looks good. Am gonna go over it again when I get home and merge then. :smile: Nice work, much appreciated.

**Comment by hyperb1iss** [comment]
Fixed an issue where the starlight sysfs node wasn't cleaned up on disconnect.

**Comment by hyperb1iss** [comment]
Got another patch to fix the massive number of wakeups from the daemon, will do a separate pull request for that.

**Review by terricain — APPROVED** [review]

**Pull request merged**